### ISSUE TITLE
Integrate with gettext to weave in translated text

### ISSUE THREAD

**Issue by mojavelinux** [asciidoctor/asciidoctor#788]
Asciidoctor should be able to process translations managed using gettext. It would use a library like [fast_gettext](https://github.com/grosser/fast_gettext) to weave the translated into the document tree prior to rendering.

Refer to the Neo4j project for some prior art in solving this problem.

**Comment by aeng** [comment]
@mojavelinux any updates on this integration?

**Comment by mojavelinux** [comment]
Nope, not yet. Unfortunately, I can't prioritize this issue right now. However, this could be something that is explored using the extension API, which came into existence after this issue was created.

**Comment by DavidGamba** [comment]
@mojavelinux Could you please describe in some detail how you envision this feature being implemented using the extension API? Right now I have no clue how to start but really need this and would like to take a stab at it.

**Comment by mojavelinux** [comment]
@DavidGamba You are in luck! Someone from the community just started working on an extension that essentially shows how this might be done. He happens to be wrapping the text in an Angular filter to allow Angular to handle the translation, but the process is the same.

See https://gist.github.com/jaredmcateer/9604e71e3eb3b39cd3285b67fb3578c3

In general, the process is as follows:

* Parse the document into a block-level AST
* Traverse the blocks and replace the text with the translated strings
* Convert the document to the target format

(I recommend using at least 1.5.6, which is still in development at the time of this comment).

There are some places where the API is missing for replacing the text. However, as that gist demonstrates, it's still possible to access the low-level instance variables that hold the source text.

There's one thing that's been bothering me a bit about how this approach might work with gettext. It would mean the translations strings would need to include all the inline AsciiDoc markup. That's actually how gettext works with DocBook (but it still feels like we could go deeper). However, I wouldn't let that stop us from getting a start on it.

**Comment by mojavelinux** [comment]
And don't forget that it's possible to translate most of the built-in labels using AsciiDoc attributes. See http://asciidoctor.org/docs/user-manual/#customizing-labels.

**Comment by martijnthe** [comment]
Thanks for asciidoctor! It's been a pleasure to use and work with.

Sharing my journey of translating docs with asciidoctor + gettext:

So far I've been using [po4a] to extract strings from `.adoc` files into `.po` files and also to generate translated `.adoc` files. It's working OK-ish. The problem is that [po4a] implements it's own "parser" (really just a bunch of regexes..) and is pretty buggy.

I started writing a replacement for `po4a-gettextize`, pretty much based upon the [design doc] that @mojavelinux wrote up 3 years ago: https://github.com/martijnthe/asciidoctor-gettext

The text extraction side of the tool is pretty complete and is working in my reasonably complex documentation project (much better than po4a so far...).

Things I ran into that I think need to be given some thought when tackling i18n as part of the core project:
- inline macros: right now the inline macro code appears in the `.po`, which is not ideal because translators may accidentally modify them. Better would be if the AST would have dedicated objects for the inline nodes.
- attributes: in the context of my project, I'd like to have the attribute references (i.e. `{some_variable}`) get emitted in the `.po` messages as-is (not substituted). In my opinion, substituting them in the `.po` output would be less DRY. The downside is that some attribute values may not be defined in an `.adoc` at all (but defined through a CLI argument) and may get missed in the translation process.
- conditional directives: I noticed that conditionals are evaluated pretty early in the parsing process and that conditionals themselves, nor the "hidden" content does not end up in the AST. This is a problem when you are trying to extract all the strings. I ended up working around this using a [preprocessor that strips out all conditionals first].
- include directives: my project uses includes quite heavily to "paste" repeated content snippets in various places while keeping the documentation sources DRY. I didn't want to get the strings from these snippets to get included in the `.po` files of the `.adoc`s that include them, so I had to create [a no-op include processor], to avoid getting duplicated strings to translate.

Next step for me is to figure out the "translation" part: injecting a `.po` back to generate a translated `.adoc` file. Ideally, I'd need #1793 to accomplish that.

I guess as an alternative approach, I could run asciidoctor with a plugin, passing it both an `.adoc` and `.po` and getting a translated `.html` back (this project also seems to take that approach: https://github.com/seikichi/asciidoctor-i18n).

Any guidance and/or tips are appreciated!

[po4a]: https://github.com/mquinson/po4a
[design doc]: https://github.com/asciidoctor/asciidoctor/wiki/Design-Doc:-i18n
[preprocessor that strips out all conditionals first]: https://github.com/martijnthe/asciidoctor-gettext/blob/master/conditionals.ts
[a no-op include processor]: https://github.com/martijnthe/asciidoctor-gettext/blob/master/includes.ts

**Comment by ciampix** [comment]
May I ask where po4a is buggy? Because we ([KiCad documentation project](http://kicad-pcb.org/help/documentation/), [(github repo)](https://github.com/kicad/kicad-doc)) are using it, and we found it already very good indeed!

**Comment by martijnthe** [comment]
@ciampix:

> May I ask where po4a is buggy? 

I had issues with conditionals (which I had fixed: https://github.com/mquinson/po4a/commit/5008a5460872edcc7649ab1fe176b1656df9bd48). Later I ran into more issues with tables and other things, causing me to give up on it.

**Comment by martijnthe** [comment]
> Next step for me is to figure out the "translation" part: injecting a .po back to generate a translated .adoc file. Ideally, I'd need #1793 to accomplish that.

I spent some time last week implementing the translator part and just published the update tool to npm: https://www.npmjs.com/package/asciidoctor-gettext

I went down the route of writing the code to take the tree of blocks and rewrite it back to an asciidoc document. It was pretty painful and hacky to get certain stuff to work with the current asciidoctor.js API. I already outlined the reasons / missing things in my earlier comment:
https://github.com/asciidoctor/asciidoctor/issues/788#issuecomment-385208899

I wish the "AST" was lower level. A lof of information from the original adoc is lost, such as conditionals, includes and comments. Includes and conditionals were particularly tricky to rewrite because they don't get emitted as nodes in the tree.
